Sources/Core/UserData.swift

```diff
@@ -0,0 +1,179 @@
+ analytics and app events.
+ */
+public struct UserData: Codable {
+  var email: String?
+  var firstName: String?
+  var lastName: String?
+  var phone: String?
+  var dateOfBirth: String?
+  var gender: String?
+  var city: String?
+  var state: String?
+  var zip: String?
+  var country: String?
+
+  init(
+    email: String? = nil,
+    firstName: String? = nil,
+    lastName: String? = nil,
+    phone: String? = nil,
+    dateOfBirth: String? = nil,
+    gender: String? = nil,
+    city: String? = nil,
+    state: String? = nil,
+    zip: String? = nil,
+    country: String? = nil
+    ) {
+    self.email = email
+    self.firstName = firstName
+    self.lastName = lastName
+    self.phone = phone
+    self.dateOfBirth = dateOfBirth
+    self.gender = gender
+    self.city = city
+    self.state = state
+    self.zip = zip
+    self.country = country
+  }
+
+  /**
+   Encodes the UserData after normalizing and hashing the information
+   */
+  public func encode(to encoder: Encoder) throws {
+    var container = encoder.container(keyedBy: CodingKeys.self)
+
+    if let email = UserData.encodable(email, forKey: .email) {
+      try container.encode(email, forKey: .email)
+    }
+
+    if let firstName = UserData.encodable(firstName, forKey: .firstName) {
+      try container.encode(firstName, forKey: .firstName)
+    }
+
+    if let lastName = UserData.encodable(lastName, forKey: .lastName) {
+      try container.encode(lastName, forKey: .lastName)
+    }
+
+    if let dateOfBirth = UserData.encodable(dateOfBirth, forKey: .dateOfBirth) {
+      try container.encode(dateOfBirth, forKey: .dateOfBirth)
+    }
+
+    if let city = UserData.encodable(city, forKey: .city) {
+      try container.encode(city, forKey: .city)
+    }
+
+    if let state = UserData.encodable(state, forKey: .state) {
+      try container.encode(state, forKey: .state)
+    }
+
+    if let country = UserData.encodable(country, forKey: .country) {
+      try container.encode(country, forKey: .country)
+    }
+
+    if let zip = UserData.encodable(zip, forKey: .zip) {
+      try container.encode(zip, forKey: .zip)
+    }
+
+    if let phone = UserData.encodable(phone, forKey: .phone) {
+      try container.encode(phone, forKey: .phone)
+    }
+
+    if let gender = UserData.encodable(gender, forKey: .gender) {
+      try container.encode(gender, forKey: .gender)
+    }
+  }
+
+  private static func encodable(_ value: String?, forKey key: CodingKeys) -> String? {
+    guard let value = value else {
+      return nil
+    }
+
+    guard !isHashed(value) else {
+      return value
+    }
+
+    let normalized = UserData.normalized(value: value, forKey: key)
+
+    guard let data = normalized.data(using: .utf8) else {
+      return nil
+    }
+    return hashed(data)
+  }
+
+  private static func isHashed(_ value: String) -> Bool {
+    return value.range(of: "[A-Fa-f0-9]{64}", options: .regularExpression) != nil
```

Comment:
If the length of string is not 64, it's not SHA256 hashed.

Reply:
Oops, didn’t mean to hit resolve on that, I’m pretty sure this will return true if the length is 64 characters. Did I understand the comment correctly?